### DISCUSSION TITLE
How to get android context for starting an intent?

### DISCUSSION THREAD

**Discussion by malliaridis** [arkivanov/Decompose#124] [Q&A] [answered]
I am currently developing a simple application with Decompose and Compose MPP. For this application, I try to implement user authentication with OAuth 2.0 and the device token flow.

This requires to open a URL in the browser of the device. For that I have provided a platform specific function. I could figure out the implementation for `desktop` and `js`, but for `android` I need a context to start an intent.

Where can I get that context from? And is the store provider the right location to call that function if following the structure of the [Todo App](https://github.com/JetBrains/compose-jb/tree/master/examples/todoapp)?

Here is the current implementation so far:
```kotlin
// commonMain
expect fun openUrlInBrowser(url: String)

// desktopMain
import java.awt.Desktop
import java.net.URI

actual fun openUrlInBrowser(url: String) {
  Desktop.getDesktop().browse(URI(url))
}

// jsMain
import kotlinx.browser.window

actual fun openUrlInBrowser(url: String) = window.open(url = url)

// androidMain
import android.content.Intent
import android.net.Uri

actual fun openUrlInBrowser(url: String) {
  val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
  `missing-context`.startActivity(intent)
}
```

_P.S. Your work is fantastic Arkadii, I could learn a lot about abstraction and structuring projects and I can finally test my components properly. Thank you very much for that._

**Comment by arkivanov** [comment, marked as answer]
If you just need to abstract the Android context, then instead of using the function directly, you can pass an interface into the component. Then implement this interface for each platform, on Android the implementing class should receive the context via constructor.

Another way would be to just pass a callback into the component - e.g. `startOauth: () -> Unit`. Then implement the callback in the activity.

**Reply by malliaridis** [reply]
I see. I was thinking of making use of the store's labels and observing those in the UI, so that the function can be annotated with `@Composable` and be called directly from inside a composable. That way I would have access to `LocalContext.current` on Android's implementation.

Would that be a good alternative?

**Reply by arkivanov** [reply]
Yes, this should also work, if calling the activity from UI is good for your architecture.

**Reply by malliaridis** [reply]
Thanks for the answer. I will try both options and see how well they work out and follow the separation of concerns principle. :)
